File: android/src/main/kotlin/com/mytiki/sdk/capture/receipt/capacitor/receipt/model/ModelSurvey.kt
/*
 * Copyright (c) TIKI Inc.
 * MIT license. See LICENSE file in root directory.
 */

package com.mytiki.sdk.capture.receipt.capacitor.receipt.model

import com.getcapacitor.JSObject
import com.microblink.core.Survey
import org.json.JSONArray
import org.json.JSONObject

/**
 * Represents a survey response for the TIKI SDK.
 *
 * @param survey The [Survey] object containing survey data.
 */
class ModelSurvey(survey: Survey) {

    /**
     * The client's user ID for the survey.
     */
    private val clientUserId: String?

    /**
     * The server ID of the survey.
     */
    private val serverId: Int

    /**
     * The slug associated with the survey.
     */
    private val slug: String?

    /**
     * The reward value associated with the survey.
     */
    private val rewardValue: String?

    /**
     * The start date of the survey.
     */
    private val startDate: String?

    /**
     * The end date of the survey.
     */
    private val endDate: String?

    /**
     * The list of survey questions.
     */
    private val questions: List<ModelSurveyQuestion>

    init {
        clientUserId = survey.clientUserId()
        serverId = survey.serverId()
        slug = survey.slug()
        rewardValue = survey.rewardValue()
        startDate = survey.startDate()
        endDate = survey.endDate()
        questions =
            survey.questions()?.map { question -> ModelSurveyQuestion(question) } ?: emptyList()
    }

    /**
     * Converts the survey response to a JSON representation.
     *
     * @return A [JSONObject] containing the survey response data.
     */
    fun toJS(): JSObject =
        JSObject()
            .put("clientUserId", clientUserId)
            .put("serverId", serverId)
            .put("slug", slug)
            .put("rewardValue", rewardValue)
            .put("startDate", startDate)
            .put("endDate", endDate)
            .put("questions", JSONArray(questions.map { question -> question.toJS() }))

    companion object {

        /**
         * Creates an instance of [ModelSurvey] from a [Survey] object.
         *
         * @param survey The [Survey] object to convert.
         * @return An [ModelSurvey] object representing the survey response, or `null` if the input is `null`.
         */
        fun opt(survey: Survey?): ModelSurvey? =
            if (survey != null) ModelSurvey(survey) else null
    }
}

File: ios/Plugin/Receipt/Model/ModelSurvey.swift
//
//  JSSurvey.swift
//  Plugin
//
//  Created by Michael Audi on 10/13/23.
//  Copyright © 2023 Max Lynch. All rights reserved.
//

import Foundation
import BlinkReceipt
import BlinkEReceipt
import Capacitor


/**
 Represents a response containing survey information.

 This struct is used to convey details about a survey, including its dates, questions, and rewards.
 */
struct ModelSurvey {
    /// The slug associated with the survey.
    private let slug: String?
    /// The reward value associated with the survey (float formated to 2 decimals)
    private let rewardValue: String?
    /// The start date of the survey (ISO8601).
    private let startDate: String?
    /// The end date of the survey (ISO8601).
    private let endDate: String?
    /// The list of survey questions.
    private let questions: [ModelSurveyQuestion]

    /**
     Initializes an `RspShipment` struct.

     - Parameters:
        - shipment: The shipment information.
     */
    init(survey: BRSurvey) {
        slug = survey.slug
        rewardValue = String(format:"%.2f", survey.rewardValue)
        startDate = survey.startDate != nil ? survey.startDate!.ISOStringFromDate() : nil
        endDate = survey.endDate != nil ? survey.endDate!.ISOStringFromDate() : nil
        questions = survey.questions.map { question in ModelSurveyQuestion(surveyQuestion: question) }
    }
    
    static func opt(survey: BRSurvey?) -> JSSurvey? {
        return survey != nil ? JSSurvey(survey: survey!) : nil
    }

    /**
     Converts the `JSSurvey` struct into a dictionary suitable for use in plugin response data.

     - Returns: A dictionary containing shipment information in a format suitable for a Capacitor plugin call result.
     */
    func toJSObject() -> JSObject {
        var ret = JSObject()
        ret["slug"] = slug
        ret["rewardValue"] = rewardValue
        ret["startDate"] = startDate
        ret["endDate"] = endDate
        ret["questions"] = JSArray(arrayLiteral: questions.map { question in question.toJSObject() })
        return ret
    }
}
